### DISCUSSION TITLE
IM7 convert is slower than IM6.

### DISCUSSION THREAD

**Discussion by Flash-nDie** [ImageMagick/ImageMagick#4637] [Help]
Hi. 
I have a question about the performance between IM7 and IM6.

My team used IM6 in thumb service until now and we are going to use IM7 in the near future. 
The most job in our service is reproduce thumb image from source image by Wand(resize, quality adjustin, changing format...)

The problem is that the conversion animated image by IM7 is slower than by IM6. 
(Read Image -> Coalesce -> Optimize -> Quantize -> Image To Blob) is main process in our service. 
Example with Wand is below.

```csharp
void test_performance(){
    MagickWand *wand = NULL;
    MagickWand *coal_wand = NULL;
    MagickWand *opt_wand = NULL;
    char *buf = NULL;
    wand = NewMagickWand();
    size_t len;
    struct timeval start;
    struct timeval end;
    unsigned long long diff;

    /* Read Image */
    gettimeofday(&start, NULL);
    MagickReadImage(wand, "/Users/kakao/Downloads/test2.gif");
    gettimeofday(&end, NULL);
    diff = (end.tv_sec - start.tv_sec)*1000 + ((end.tv_usec - start.tv_usec)/1000);
    printf("Read Image : %lld\n", diff);

    /* Resize */
    gettimeofday(&start, NULL);
    MagickResetIterator(wand);
    for (int i = 0; MagickNextImage(wand) != MagickFalse; i++) {
        // Image resizing
#if MAGICK_VERSION == 7
        if (MagickResizeImage(wand, 480, 480, LanczosFilter) != MagickTrue) {
#else
            if (MagickResizeImage(wand, 480, 480, LanczosFilter, 1) != MagickTrue) {
#endif
            printf("resize error\n");
        }
    }
    gettimeofday(&end, NULL);
    diff = (end.tv_sec - start.tv_sec)*1000 + ((end.tv_usec - start.tv_usec)/1000);
    printf("Resize Image : %lld\n", diff);
    MagickResetIterator(wand);

    /* coalesce */
    gettimeofday(&start, NULL);
    if ((coal_wand = MagickCoalesceImages(wand)) == NULL) {
        printf("coalesce error\n");
    }
    gettimeofday(&end, NULL);
    diff = (end.tv_sec - start.tv_sec)*1000 + ((end.tv_usec - start.tv_usec)/1000);
    printf("Coalesce Image : %lld\n", diff);
    DestroyMagickWand(wand);
    wand = coal_wand;

    /* optimize */
    gettimeofday(&start, NULL);
    if ((opt_wand = MagickOptimizeImageLayers(wand)) == NULL) {
        printf("optimize error\n");
    }
    gettimeofday(&end, NULL);
    diff = (end.tv_sec - start.tv_sec)*1000 + ((end.tv_usec - start.tv_usec)/1000);
    printf("Optimize Image : %lld\n", diff);
    DestroyMagickWand(wand);
    wand = opt_wand;

    /* quantize */
    gettimeofday(&start, NULL);
    if (MagickQuantizeImages(wand, 255, RGBColorspace, 0, MagickFalse, MagickFalse) == MagickFalse) {
        printf("quantize error\n");
    }
    gettimeofday(&end, NULL);
    diff = (end.tv_sec - start.tv_sec)*1000 + ((end.tv_usec - start.tv_usec)/1000);
    printf("Quantize Image : %lld\n", diff);

    /* image to blob */
    gettimeofday(&start, NULL);
    buf = MagickGetImagesBlob(wand, &len);
    gettimeofday(&end, NULL);
    diff = (end.tv_sec - start.tv_sec)*1000 + ((end.tv_usec - start.tv_usec)/1000);
    printf("Image to Blob : %lld\n", diff);

}
```

When checking processing time, the overall steps in logic spent more time in IM7, especially in Optimize step. 
I tested some animated gif and animated webp, and those show same results that IM7 spent longer. 

I just tested with CLI convert like this : convert test.webp -resize 480x -coalesce -layers optimize -colors 255 out6.webp
With CLI, IM7 is slower than IM6 by 30~50%. 

Is there some difference in IM6 and IM7 making performance worse? I want to know some ways to get better performance with IM7. 

Below is my environment
1. OS 
i tested in macOS Catalina 10.15.7 and CentOS 7

2. IM6, IM7 version and configure
IM6
    version : 6.9.11-31
    configure :   --disable-shared --disable-openmp --without-perl --without-fftw --without-magick-plus-plus --without-x --without-pango --without-djvu --with-lcms --without-lqr --without-openexr --without-openjp2 --with-webp --with-xml --with-heic --with-rsvg
IM7
    version : 7.1.0-17-2
    configure :   --disable-shared --disable-openmp --without-perl --without-fftw --without-magick-plus-plus --without-x --without-pango --without-djvu --with-lcms --without-lqr --without-openexr --without-openjp2 --with-webp --with-xml --with-heic --with-rsvg --enable-hdri=no

* some test files are attached 
![test](https://user-images.githubusercontent.com/67739935/147643230-4b594a30-e8a5-4d9e-9615-3278108b0f70.gif)
![test2](https://user-images.githubusercontent.com/67739935/147643246-abd7fd55-eb6a-4432-af2f-db67c352929c.gif)

**Comment by snibgo** [comment]
IM v7 has an extra level of indirection when accessing pixel values. This might account for the decreased performance.

I assume both versions are Q16. As you are creating GIFs, which have only 8 bits per channel, you might use a Q8 build for increased performance.

**Comment by Flash-nDie** [comment]
Yes. I used Q16. 
1. But We deal with other type(webp, jpg, png, etc...), So we uses Q16. 

2. Building with --with-quantum-depth=8 show similar results(processing time is slightly increased)

Is there any other way to increase Performance? Any configure options or additional API according to architecture of IM7.

**Comment by snibgo** [comment]
The order of your operations seem strange. If "-coalesce" does anything, this is because some frames are smaller than others, containing just pixels that have changed. But if so, then resizing all frames to the same size before coalescing won't do what you want.

So I suspect your "-coalesce" does nothing.

I suspect that optimizing the layers is more effective if they are already reduced to 255 colours.

**Comment by urban-warrior** [comment]
Optimize is much slower in IMv7.  We will investigate but we will need a few weeks to resolve this issue.

**Comment by Flash-nDie** [comment]
@snibgo 
Thank you for modifying CLI sequence. Unlike the test code, I used resizes before coalesce. 
I tested two sequence. 
1. -coalesce -resize 480x -layers optimize -colors 255 (resize after coalesce)
2. -coalesce -resize 480x  -colors 255 -layers optimize (as you mentioned)

but still two cases is slower in IM7. 
urban-warrior says that investigation about Optimize. 
Thank you for dealing with discussion!

@urban-warrior 
Thank you for answer. 
Before investigation, I saw the processing time in IM7 is 20~30 times slower than in IM6.(of course, it depends on each cases...)
So I will wait your investigation. 
One more things, the Image loading(MagickReadImages, MagickReadImageBlob, etc..) also is slower than in IM6. 
Is it caused by extra level of indirection in IM7 mentioned by snibgo? 

> IM v7 has an extra level of indirection when accessing pixel values. This might account for the decreased performance.

Anyway, How long will the investigation take?

**Comment by urban-warrior** [comment]
ImageMagick version 7 supports variable channels and improved masking as compared to ImageMagick version 6.  See https://imagemagick.org/script/porting.php.  This flexibility does come with a performance hit as we loop over an image in scanline order we have internal looping over the channels as well as checks for read and write masks.  The increased branching inside the loops is the likely source of the performance issue.  Your script returned 126 for reading the blocks image for IMv6 whereas its 137 for IMv7 which is marginal.   The performance hit is primarily with the image optimization and some with the quantization.  

When will we investigate?  Don't know.  We're working on seamless compositing of two images.  The time it takes is variable.  It could be a few days or a few weeks.

**Comment by fmw42** [comment]
Recompile without HDRI might improve performance.

**Comment by Flash-nDie** [comment]
@urban-warrior 
Thank you for answering.
We are waiting for you to complete the task.

@fmw42 
Hi. IM7 configures below. We compiled IM7 without HDRI already. 
> version : 7.1.0-17-2
configure : --disable-shared --disable-openmp --without-perl --without-fftw --without-magick-plus-plus --without-x --without-pango --without-djvu --with-lcms --without-lqr --without-openexr --without-openjp2 --with-webp --with-xml --with-heic --with-rsvg --enable-hdri=no

**Comment by Jmuccigr** [comment]
For the CLI, I'm curious whether running `magick` in IM 7 instead of `convert` makes any difference.

**Comment by urban-warrior** [comment]
Not in performance but certainly how the command-line is parsed.  See the porting guide for details.  In general, you should use `magick` rather than `magick convert`.

**Comment by urban-warrior** [comment]
ImageMagick 7 had numerous bug fixes and enhancements as compared to IMv6 per the [porting guide](https://imagemagick.org/script/porting.php).  One bug fix is comparing the image bounding box when optimizing layers.  There is a use case that fails in IMv6 but is captured in IMv7.  Because of the missing use case in IMv6, there is an opportunity to exit the algorithm early whereas in IMv7, the algorithm scans the entire image each iteration.  That likely explains the extended optimization time.  Otherwise, any reduction in processing time is likely related to variable pixel channels.  As mentioned, in IMv7, we iterate by rows then columns then channels.  Channels are mapped as they can appear in any order. For IMv6, the channels are fixed at 4 quantums in predictable places and iterations are rows by columns.   In addition, IMv7 supports a read and write mask channels.  Image processing operators in IMv7 often check the channel to see if a particular pixel is masked.  This additional branching in a loop is an additional performance hit.  However, per the porting guide, the benefits of the enhancements/bug fixes/variable channels typically outweigh the performance hit-- but not always.

**Comment by Flash-nDie** [comment]
In summary, are you saying that the reason the processing time increases during the optimization process is not a bug to be fixed, but a natural result after fixing a bug in IM6?
If it is true, it is questionable that the optimization execution time in IM7 differs by 20 to 30 times compared to IM6, even considering  bug fixes and structural differences.

**Comment by urban-warrior** [comment]
IMv7 takes a performance hit when compared to IMv6 for the reasons stated.  In most cases its negligible, however, for algorithms that iterate, there can be an accumulative effect that is noticeable.   Your image iterates ~200 times when computing the bounding box and reducing colors to add an extra ~2 seconds to your run time.  The solution may be to stick with IMv6 for this use case.

Here are the results of the timing per your MagickWand app:
```
IMv7

$./im-performance banana.gif
Read Image : 111
Resize Image : 226
Coalesce Image : 25
Optimize Image : 115
Quantize Image : 194
Image to Blob : 9
$ ./im-performance blocks.gif
Read Image : 141
Resize Image : 495
Coalesce Image : 132
Optimize Image : 1059
Quantize Image : 3416
Image to Blob : 290

IMv6

$ ./im-performance banana.gif
Read Image : 131
Resize Image : 128
Coalesce Image : 15
Optimize Image : 19
Quantize Image : 39
Image to Blob : 6
echo "  blocks"
  blocks
$ ./im-performance blocks.gif
Read Image : 121
Resize Image : 301
Coalesce Image : 127
Optimize Image : 100
Quantize Image : 2931
Image to Blob : 23
```
